In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 16: Aunt Sue ---</h2><p>Your Aunt Sue has given you a wonderful gift, and you'd like to send her a thank you card.  However, there's a small problem: she signed it "From, Aunt Sue".</p>
<p>You have 500 Aunts named "Sue".</p>
<p>So, to avoid sending the card to the wrong person, you need to figure out which Aunt Sue (which you conveniently number 1 to 500, for sanity) gave you the gift.  You open the present and, as luck would have it, good ol' Aunt Sue got you a My First Crime Scene Analysis Machine!  Just what you wanted.  Or needed, as the case may be.</p>
<p>The My First Crime Scene Analysis Machine (MFCSAM for short) can detect a few specific compounds in a given sample, as well as how many distinct kinds of those compounds there are. According to the instructions, these are what the MFCSAM can detect:</p>
<ul>
<li><code>children</code>, by human DNA age analysis.</li>
<li><code>cats</code>.  It doesn't differentiate individual breeds.</li>
<li>Several <span title="It can tell them apart by their distinct Dog Residue.">seemingly random breeds of dog</span>: <code><a href="https://en.wikipedia.org/wiki/Samoyed_%28dog%29">samoyeds</a></code>, <code><a href="https://en.wikipedia.org/wiki/Pomeranian_%28dog%29">pomeranians</a></code>, <code><a href="https://en.wikipedia.org/wiki/Akita_%28dog%29">akitas</a></code>, and <code><a href="https://en.wikipedia.org/wiki/Vizsla">vizslas</a></code>.</li>
<li><code>goldfish</code>.  No other kinds of fish.</li>
<li><code>trees</code>, all in one group.</li>
<li><code>cars</code>, presumably by exhaust or gasoline or something.</li>
<li><code>perfumes</code>, which is handy, since many of your Aunts Sue wear a few kinds.</li>
</ul>
<p>In fact, many of your Aunts Sue have many of these.  You put the wrapping from the gift into the MFCSAM.  It beeps inquisitively at you a few times and then prints out a message on <a href="https://en.wikipedia.org/wiki/Ticker_tape">ticker tape</a>:</p>
<pre><code>children: 3
cats: 7
samoyeds: 2
pomeranians: 3
akitas: 0
vizslas: 0
goldfish: 5
trees: 3
cars: 2
perfumes: 1
</code></pre>
<p>You make a list of the things you can remember about each Aunt Sue.  Things missing from your list aren't zero - you simply don't remember the value.</p>
<p>What is the <em>number</em> of the Sue that got you the gift?</p>
</article>


In [83]:
from pprint import pprint


tickertape = """
children: 3
cats: 7
samoyeds: 2
pomeranians: 3
akitas: 0
vizslas: 0
goldfish: 5
trees: 3
cars: 2
perfumes: 1
"""


def parse_ticker(tickertape: str) -> dict[str, int]:
    given = {}
    for l in tickertape.strip().splitlines():
        name, i = l.split(": ")
        given[name] = int(i)
    return given


def parse_sues(s: str) -> dict[str, dict[str, int]]:
    sues = {}
    for l in s.strip().splitlines():
        i = re.search(r"Sue \d+:\s+", l)
        sue = l[: i.end() - 2]
        things = l[i.end() :]
        sues[sue] = {}
        sue = sues[sue]
        for thing in things.split(", "):
            name, count = thing.split(": ")
            sue[name] = int(count)

    return sues


ticker = parse_ticker(tickertape)

with open("../input/day16.txt") as f:
    sues = parse_sues(f.read())


def could_be_the_sue(ticker: dict[str, int], sue: dict[str, int]) -> bool:
    for name, value in sue.items():
        if value != ticker[name]:
            return False
    return True


next(sue for sue in sues.keys() if could_be_the_sue(ticker, sues[sue]))

'Sue 373'

<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>373</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>As you're about to send the thank you note, something in the MFCSAM's instructions catches your eye.  Apparently, it has an outdated <a href="https://www.youtube.com/watch?v=RXJKdh1KZ0w">retroencabulator</a>, and so the output from the machine isn't exact values - some of them indicate ranges.</p>
<p>In particular, the <code>cats</code> and <code>trees</code> readings indicates that there are <em>greater than</em> that many (due to the unpredictable nuclear decay of cat dander and tree pollen), while the <code>pomeranians</code> and <code>goldfish</code> readings indicate that there are <em>fewer than</em> that many (due to the modial interaction of magnetoreluctance).</p>
<p>What is the <em>number</em> of the real Aunt Sue?</p>
</article>

</main>


In [86]:
def could_be_the_sue_II(ticker: dict[str, int], sue: dict[str, int]) -> bool:
    for name, value in sue.items():
        if name in ("cats", "trees"):
            if value <= ticker[name]:
                return False
        elif name in ("pomeranians", "goldfish"):
            if value >= ticker[name]:
                return False
        elif value != ticker[name]:
            return False
    return True


next(sue for sue, vs in sues.items() if could_be_the_sue_II(ticker, vs))

'Sue 260'

<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>260</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="16/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Aunt+Sue%22+%2D+Day+16+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F16&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Aunt+Sue%22+%2D+Day+16+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F16';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
